x-pack/plugin/downsample/src/main/java/org/elasticsearch/xpack/downsample/LabelFieldProducer.java

```diff
@@ -20,152 +20,139 @@
 import java.util.List;
 
 /**
- * Class that produces values for a label field.
+ * Class that produces the last value of a field for downsampling.
+ * Important note: This class assumes that field values are collected and sorted by descending order by time
  */
-abstract class LabelFieldProducer extends AbstractDownsampleFieldProducer {
+class LabelFieldProducer extends AbstractDownsampleFieldProducer {
+    // When downsampling metrics, we only keep one value even if the field was a multi-vlaue field.
+    // For labels, we preserve all the values of the last occurence.
+    private final boolean supportsMultiValue;
+    Object lastValue = null;
 
-    LabelFieldProducer(String name) {
+    LabelFieldProducer(String name, boolean producesMultiValue) {
         super(name);
+        this.supportsMultiValue = producesMultiValue;
     }
 
-    abstract Label label();
-
-    abstract static sealed class Label {
-        private final String name;
-
-        /**
-         * Abstract class that defines how a label is downsampled.
-         * @param name the name of the field as it will be stored in the downsampled document
-         */
-        protected Label(String name) {
-            this.name = name;
-        }
+    /**
+     * Creates a producer that can be used for downsampling labels. It works for all types apart from
+     * `aggregate_metric_double`, if the field type is aggregate metric double, please use
+     * {@link LabelFieldProducer#createForAggregateSubMetricLabel(String, Metric)}.
+     */
+    static LabelFieldProducer createForLabel(String name, String fieldType) {
+        assert "aggregate_metric_double".equals(fieldType) == false
+            : "field type cannot be aggregate metric double: " + fieldType + " for field " + name;
+        if ("histogram".equals(fieldType)) {
+            return new LabelFieldProducer.HistogramFieldProducer(name, true);
+        } else if ("flattened".equals(fieldType)) {
+            return new LabelFieldProducer.FlattenedFieldProducer(name, true);
+        }
+        return new LabelFieldProducer(name, true);
+    }
 
-        public String name() {
-            return name;
-        }
+    /**
+     * Creates a producer that can be used for downsampling labels. It works for all types apart from
+     * `aggregate_metric_double`, if the field type is aggregate metric double, please use
+     * {@link LabelFieldProducer#createForAggregateSubMetricMetric(String, Metric)} (String, Metric).}
+     */
+    static LabelFieldProducer createForMetric(String name) {
+        return new LabelFieldProducer(name, false);
```

Comment:
I think this makes sense and shouldn't have a negative impact on performance because only value per tsid / time bucket is read. Just to be sure we should keep an eye on the nightly tsdb benchmark after this change has been merged.